### PR TITLE
Add documentation about podvm build process

### PR DESCRIPTION
**Pull request by wainersm**
While working on https://github.com/confidential-containers/cloud-api-adaptor/pull/345 we realized the process to build podvm within container is not documented, can be confusing and demands memory/disk from the host. So the main goal of this PR is to document the process to build the podvm with or without containers.

And while working on the podvm/README.md  I realized that:

 * we don't document how to consume the pre-built images either, so I add a new article in docs.
 * we could have a script to ease the process to extract the qcow2 file from the container image, so I created a script. A few minutes about I noticed @jtumber-ibm  had the same idea (https://github.com/confidential-containers/cloud-api-adaptor/issues/563) and his script (https://github.com/confidential-containers/cloud-api-adaptor/pull/565/commits/de6725588e474ffeb9853d08fec53b7ba6244e76) is better actually :)

I didn't document the tricks to build the RHEL image because I never built it and I actually I lack time to try. Maybe @snir911 and/or @whitebrandy could help here? :)

### CODE REVIEW

**Comment by wainersm** [comment]
hi @jtumber-ibm ,

I tried to build the s390x image on my x86_64 laptop just like I documented on the podvm/README.md but it failed on the packer step. Did I miss something?

Also, I am okay to drop my extract-qcow2.sh in favor of your script. I will just need to ensure it is merged before and I can change the docs.

File: docs/consuming-prebuilt-podvm-images.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+You can find the images available at https://quay.io/organization/confidential-containers with the *podvm-[CLOUD_PROVIDER]-[DISTRO]* name pattern. For example, https://quay.io/repository/confidential-containers/podvm-libvirt-ubuntu hosts the Ubuntu images for Libvirt.
+
+The easiest way to extract the qcow2 file from the podvm container image is using the [`extract-qcow2.sh`](../podvm/hack/extract-qcow2.sh) script. For example, to extract the file from the *podvm-libvirt-ubuntu* image:
+
+```
+$ cd podvm
+$ ./hack/extract-qcow2.sh quay.io/confidential-containers/podvm-libvirt-ubuntu
+```
+
+If you need to customize the image, for example, to add a [registries authentication configuration](registries-authentication.md) then it can be done with the [libguestfs](https://www.libguestfs.org) tools. Alternatively, you can find instructions on [podvm/README.md](../podvm/README.md) on how to build the image from sources.
```
Comment (snir911) [inline docs/consuming-prebuilt-podvm-images.md:14]: can you use [statically embedded registries authentication configuration](registries-authentication.md#statically-embed-authentication-file-in-podvm-image) instead? as it's not needed if auth file passed at runtime
raw:
`[statically embedded registries authentication configuration](registries-authentication.md#statically-embed-authentication-file-in-podvm-image)`

File: podvm/README.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
+# Introduction
+
+This directory contains the sources to build the podvm image (qcow2 file) for various Linux distributions and cloud providers. So use
+the instructions on the next sections if you need to build your own image with changes to meet your requirements. Otherwise you can
+find [here](../docs/consuming-prebuilt-podvm-images.md) information on how to consume pre-built images.
+
+# How to build on locally
```
Comment (snir911) [inline podvm/README.md:7]: remove "on"

File: podvm/README.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+In order to build locally you must [packer](https://www.packer.io/) and QEMU installed in your host. Apart from those, it requires the source trees and software mentioned on the [developer's guide](docs/DEVELOPMENT.md).
```
Comment (snir911) [inline podvm/README.md:9]: you must **have** ?

File: podvm/README.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+Run the following commands to build locally:
+
+```
+$ export CLOUD_PROVIDER=[aws|azure|ibmcloud|libvirt]
+$ make binaries
+$ make image
```
Comment (snir911) [inline podvm/README.md:16]: nit: make image would be enough

**Comment by jtumber-ibm** [comment]
> hi @jtumber-ibm ,
> 
> I tried to build the s390x image on my x86_64 laptop just like I documented on the podvm/README.md but it failed on the packer step. Did I miss something?
> 
> Also, I am okay to drop my extract-qcow2.sh in favor of your script. I will just need to ensure it is merged before and I can change the docs.

hi @wainersm, did you have the s390x components installed? i.e. rust target (s390x-unknown-linux-gnu), qemu (qemu-system-s390x) and gcc (gcc-s390x-linux-gnu).

Feel free to use my script, I just need a couple other IBMers to test the ibmcloud part of the script but should be able to get that merged soon or if that gets delayed I can create a separate PR for the qcow2 from image bit.

File: podvm/README.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+```
+
+# How to build within container
+
+This directory contains dockerfiles to build the podvm image entirely within container so that it only requires docker or podman installed on the host.
+
+In general it is needed to follow these steps:
+
+1. Build a builder container image for a given Linux distribution
```
Comment (snir911) [inline podvm/README.md:25]: Ideally we can have most builders publicly available, but i think we didn't push them

File: podvm/README.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+1. Extract the podvm image (qcow2 file) from the container image
+
+The next sections describe that process in details. Note that although the following examples use docker, it can be carried out with podman too.
+
+## Building a builder image
+
+The builder image packages the cloud-api-adaptor and Kata Containers sources as well as the softwares to build
+the binaries (e.g. *kata-agent* and *agent-protocol-forwarder*) that should be installed in the podvm image.
+
+The builder image is agnostic to cloud providers in the sense that one can be used to buil for multiple providers, however it is
```
Comment (snir911) [inline podvm/README.md:36]: typo: "buil"

File: podvm/README.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+dependent of the Linux distribution the image is built for. Therefore, in this directory you will find dockerfiles to each supported distributions, which are currently Ubuntu 20.04 ([Dockerfile.podvm_builder](./Dockerfile.podvm_builder)), CentOS Stream 8 ([Dockerfile.podvm_builder.centos](./Dockerfile.podvm_builder.centos)), and RHEL 8.7 ([Dockerfile.podvm_builder.rhel](./Dockerfile.podvm_builder.rhel)).
+
+As an example, to build the builder image for Ubuntu, run:
+
+```
+$ docker build -t podvm_builder -f Dockerfile.podvm_builder .
```
Comment (snir911) [inline podvm/README.md:42]: Do you know if the "." is actually needed? at least with podman it works without it

**Comment by stevenhorsman** [comment]
> > hi @jtumber-ibm ,
> > I tried to build the s390x image on my x86_64 laptop just like I documented on the podvm/README.md but it failed on the packer step. Did I miss something?
> > Also, I am okay to drop my extract-qcow2.sh in favor of your script. I will just need to ensure it is merged before and I can change the docs.
> 
> hi @wainersm, did you have the s390x components installed? i.e. rust target (s390x-unknown-linux-gnu), qemu (qemu-system-s390x) and gcc (gcc-s390x-linux-gnu).
> 
> Feel free to use my script, I just need a couple other IBMers to test the ibmcloud part of the script but should be able to get that merged soon or if that gets delayed I can create a separate PR for the qcow2 from image bit.

FYI - this is high on my list, so I hope to get to it this morning.

File: podvm/README.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+|Argument|Default value|Description|
+|--------|-------------|-----------|
+|CAA\_SRC |https://github.com/confidential-containers/cloud-api-adaptor | The cloud-api-adaptor source repository |
+|CAA\_SRC\_BRANCH|staging| cloud-api-adaptor repository branch |
+|KATA\_SRC | https://github.com/kata-containers/kata-containers | The Kata Containers source repository |
+|KATA\_SRC\_BRANCH | CCv0 | The Kata Containers repository branch |
+|GO\_VERSION | 1.18.7 | Go version |
+|PROTOC\_VERSION | 3.11.4 | [Protobuf](https://github.com/protocolbuffers/protobuf) version |
+|RUST\_VERSION | 1.66.0 | Rust version |
```
Comment (snir911) [inline podvm/README.md:56]: I think we will be able to drop this in favor of allowing to pass CAA_SRC/BRANCH and KATA_SRC/BRANCH to the build containers (and there override the repos) BUT we currently don't have this option so i guess it's fine :)

File: podvm/README.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+* Overwrite the *CAA_SRC* and *CAA_SRC_BRANCH* arguments as shown below:
+
+```
+$ docker build -t podvm_builder \
+	--build-arg CAA_SRC=https://github.com/$USER/cloud-api-adaptor \
+	--build-arg CAA_SRC_BRANCH=my-changes-in-a-branch \
+	-f Dockerfile.podvm_builder .
+```
+
+## Building the podvm container image
```
Comment (snir911) [inline podvm/README.md:71]: how about just "Building the podvm qcow2 image", below you can mentioned it's wrapped in a container image

File: podvm/README.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+add the packer configuration files.
+
+Follow the steps below, replacing `DISTRO` with the name of the distribution being added:
+
+1. Create the builder dockerfile by copying `Dockerfile.podvm_builder` to `Dockerfile.podvm_builder.DISTRO` and
+   adjusting the file properly (e.g. replace `FROM ubuntu:20.04` with `FROM DISTRO`). Try to keep the same
+   software versions (e.g. Golang, Rust) as much as possible.
+1. Create the podvm image dockerfile by copying `Dockerfile.podvm` to `Dockerfile.podvm.DISTRO` and adjusting the file
+   properly likewise. In particular, the *PODVM_DISTRO* and *BUILDER_IMG* arguments should be changed.
+1. Create the packer directory (`mkdir qcow2/DISTRO`) where the `qemu-DISTRO.pkr.hcl` and `variables.pkr.hcl` files should be placed. Also on this step you can also use an existing configuration (e.g. `qcow2/ubuntu`) as a template. Ensure that common scripts and files like the `qcow2/misc-settings.sh` are adjusted to support the DISTRO if needed.
```
Comment (snir911) [inline podvm/README.md:127]: Makefile should be modified as well

**Review by snir911 — COMMENTED** [review]
Thanks Wainer! it is super useful!, i added some comments, also i think worth pointing to it from the install instructions

**Comment by snir911** [comment]
when it comes to RHEL build you just need to run it from a subscribed system (unless pkgs needs to be installed in the packer vm then you'll need to use libguestfs to install it e.g. open-vm-tools for vsphere) cc @whitebrandy to confirm

**Review by stevenhorsman — APPROVED** [review]
This is really helpful, thanks! I wonder if we should replace some of the sections in the [libvirt doc](https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/libvirt/README.md) (and others later probably) with a link to this now?
I've tried both methods and left a bit of feedback. I didn't test extract-qcow though if it is likely to be replaced with JT's code.

**Comment by jtumber-ibm** [comment]
@wainersm [podvm/hack/download-image.sh](https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/hack/download-image.sh) merged

**Comment by wainersm** [comment]
> @wainersm [podvm/hack/download-image.sh](https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/hack/download-image.sh) merged

Cool! I will drop my script and adjust the docs. thanks @jtumber-ibm !

File: podvm/README.md (hunk 5 above)
Comment (bpradipt) [reply podvm/README.md:25]: Yeah that was the intent. Having podvm_builder images published for releases (including any pre-releases)

File: podvm/README.md (hunk 4 above)
Comment (wainersm) [reply podvm/README.md:16]: ah, cool. I will adjust that.

File: podvm/README.md (hunk 5 above)
Comment (wainersm) [reply podvm/README.md:25]: Agreed. Actually I created this README aiming at developers that need to build podvm from latest code. And the [docs/consuming-prebuilt-podvm-images.md](https://github.com/confidential-containers/cloud-api-adaptor/pull/568/files/d5dab40a507fcd557cfa9747547ee6014a77dc78#diff-4fc2fb53a97f8dc41a360dc7a32b8dda5ff90f13819ce81e06f6c4e373e9e956) target consumers of the project.

File: podvm/README.md (hunk 6 above)
Comment (wainersm) [reply podvm/README.md:36]: you really read the doc :)

File: podvm/README.md (hunk 7 above)
Comment (wainersm) [reply podvm/README.md:42]: yes, it does. At least the docker version (20.10.17) I am using:
```
$ docker build -t podvm_builder -f Dockerfile.podvm_builder
"docker build" requires exactly 1 argument.
```

File: podvm/README.md (hunk 10 above)
Comment (wainersm) [reply podvm/README.md:127]: True. thanks, I will add that.

File: podvm/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (wainersm) [reply podvm/README.md:9]: Yeah, we got dependencies to build CAA binaries, kata containers as well as the image itself.

qemu-utils, cloud-utils and qemu-kvm are likely needed by packer. Rust, musl and the other software for the kata binaries.

I will try to re-phase that to capture the mess this whole thing is :)

File: docs/consuming-prebuilt-podvm-images.md [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# Consuming pre-built podvm images
+
+The project has automation in place to build container images that contain the qcow2 file for the various Linux distributions and cloud providers supported. Those images are re-built after each pull request are merged into this repository.
```
[the remark this answers is not in the document]
Comment (wainersm) [reply docs/consuming-prebuilt-podvm-images.md]: you are right. I got it wrong that we were building the images for every pull request. I fixed that information on the version of this PR that I just pushed.

File: docs/consuming-prebuilt-podvm-images.md (hunk 1 above)
Comment (wainersm) [reply docs/consuming-prebuilt-podvm-images.md]: I am still a little confused here but I gave it a try and re-wrote that section. Let me know if it looks good (or not).

**Comment by wainersm** [comment]
With the latest updates on this PR I addressed all comments made by @snir911 , @stevenhorsman and @bpradipt . I also adjusted to reference the download-image.sh script created by @jtumber-ibm 

Pls let me know if it is good to get merged.

File: docs/consuming-prebuilt-podvm-images.md [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+You can find the images available at https://quay.io/organization/confidential-containers with the *podvm-[CLOUD_PROVIDER]-[DISTRO]* name pattern. For example, https://quay.io/repository/confidential-containers/podvm-libvirt-ubuntu hosts the Ubuntu images for Libvirt.
+
+The easiest way to extract the qcow2 file from the podvm container image is using the [`download-image.sh`](../podvm/hack/download-image.sh) script. For example, to extract the file from the *podvm-libvirt-ubuntu* image:
+
+```
+$ cd podvm
+$ ./hack/download-image.sh quay.io/confidential-containers/podvm-libvirt-ubuntu . -o podvm.qcow2
+```
+
+In case your workload images are pulled from a private registry then you need to provide the authentication file by either [installing along with the cloud-api-adaptor deployment](registries-authentication.md#deploy-authentication-file-along-with-cloud-api-adaptor-deployment) or [statically embeddeding in the podvm](registries-authentication.md#statically-embed-authentication-file-in-podvm-image). With the later you will need to build the image from sources, so find detailed instructions on [podvm/README.md](../podvm/README.md).
```
Comment (bpradipt) [inline docs/consuming-prebuilt-podvm-images.md:14]: nit: "statically embeddeding in the podvm" => "statically **embedding** in the podvm **image**"

nit: "so find detailed instructions on" => "so find detailed instructions **in**"

File: podvm/README.md [hunk 13]
```diff
@@ -0,0 +1,134 @@
+# Introduction
+
+This directory contains the sources to build the podvm image (qcow2 file) for various Linux distributions and cloud providers. So use
+the instructions on the next sections if you need to build your own image with changes to meet your requirements. Otherwise you can
```
Comment (bpradipt) [inline podvm/README.md:4]: nit: "So use the instructions on the next sections" => "So use the instructions **in** the next sections"

File: podvm/README.md [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+find [here](../docs/consuming-prebuilt-podvm-images.md) information on how to consume pre-built images.
+
+# How to build locally
+
+In order to build locally it requires the source trees and softwares mentioned on the [developer's guide](../docs/DEVELOPMENT.md) to build this project binaries. It will also need [packer](https://www.packer.io/) (to build the qcow2), [rust](https://www.rust-lang.org/tools/install) (to build the Kata Containers's agent), as well as the following packages:
```
Comment (bpradipt) [inline podvm/README.md:9]: nit: "on the [developer's guide]" => "in the [developer's guide]"

File: podvm/README.md [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+The next sections describe that process in details. Note that although the following examples use docker, it can be carried out with podman too.
+
+## Building a builder image
+
+The builder image packages the cloud-api-adaptor and Kata Containers sources as well as the softwares to build
+the binaries (e.g. *kata-agent* and *agent-protocol-forwarder*) that should be installed in the podvm image.
+
+The builder image is agnostic to cloud providers in the sense that one can be used to build for multiple providers, however it is
+dependent of the Linux distribution the image is built for. Therefore, in this directory you will find dockerfiles to each supported distributions, which are currently Ubuntu 20.04 ([Dockerfile.podvm_builder](./Dockerfile.podvm_builder)), CentOS Stream 8 ([Dockerfile.podvm_builder.centos](./Dockerfile.podvm_builder.centos)), and RHEL 8.7 ([Dockerfile.podvm_builder.rhel](./Dockerfile.podvm_builder.rhel)).
```
Comment (bpradipt) [inline podvm/README.md:42]: nit: "of the Linux" => "on the Linux"
nit: "in this directory you will find dockerfiles to each supported distributions" => "in this directory you will find dockerfiles **for** each supported distributions"

File: podvm/README.md [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+|--------|-------------|-----------|
+|CAA\_SRC |https://github.com/confidential-containers/cloud-api-adaptor | The cloud-api-adaptor source repository |
+|CAA\_SRC\_BRANCH|staging| cloud-api-adaptor repository branch |
+|KATA\_SRC | https://github.com/kata-containers/kata-containers | The Kata Containers source repository |
+|KATA\_SRC\_BRANCH | CCv0 | The Kata Containers repository branch |
+|GO\_VERSION | 1.18.7 | Go version |
+|PROTOC\_VERSION | 3.11.4 | [Protobuf](https://github.com/protocolbuffers/protobuf) version |
+|RUST\_VERSION | 1.66.0 | Rust version |
+
+As it can be noted on the table above the cloud-api-adaptor repository is cloned within the builder image, so rather than
```
Comment (bpradipt) [inline podvm/README.md:63]: nit: "on the table above" => "in the table above"

File: podvm/README.md [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+copying the local source tree, it will be used the upstream source. But if you want to test local changes then you should:
```
Comment (bpradipt) [inline podvm/README.md:64]: nit: "it will be used the upstream source" => "it will be **using** the upstream source"

File: podvm/README.md [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+
+* Push the changes to elsewhere (e.g. https://github.com/$USER/cloud-api-adaptor/tree/my-changes-in-a-branch).
```
Comment (bpradipt) [inline podvm/README.md:66]: nit: "to elsewhere" => "to your fork in github"

File: podvm/README.md [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+	--build-arg CAA_SRC=https://github.com/$USER/cloud-api-adaptor \
+	--build-arg CAA_SRC_BRANCH=my-changes-in-a-branch \
+	-f Dockerfile.podvm_builder .
+```
+
+## Building the podvm qcow2 image
+
+The process to build the podvm container image will effectively compile binaries and create the qcow2 file.
+
+In order to build the podvm image there should be used the corresponding dockerfile of the Linux distro which the builder image (see the [previous step](#building-a-builder-image)) was built. For example, if the builder image was built with *Dockerfile.podvm_builder.DISTRO* then you should use the *Dockerfile.podvm.DISTRO* to build the podvm image.
```
Comment (bpradipt) [inline podvm/README.md:80]: "In order to build the podvm image there should be used the corresponding dockerfile of the Linux distro which the builder image was built" => "In order to build the podvm image you should be using the corresponding dockerfile of the Linux distro for which the builder image was built"

File: podvm/README.md [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+
+The builder image has to be indicated via `BUILDER_IMG` build argument to docker. Apart from that, the `CLOUD_PROVIDER` argument is mandatory and specify the cloud provider that binaries should be built for. Once again, one builder image can be used to build for multiple
```
Comment (bpradipt) [inline podvm/README.md:82]: nit: "and specify the cloud provider " => "and **specifies** the cloud provider"

File: podvm/README.md [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+we have the [hack/download-image.sh](hack/download-image.sh) script, which copy the qcow2 file out of the podvm
+container image. Use it as shown below:
+
+```
+$ ./hack/download-image.sh podvm_libvirt:latest . -o podvm.qcow2
+```
+
+# How to add support for a new Linux distribution
+
+In order to add a new Linux distribution essentially it is needed to create some dockerfile files and
```
Comment (bpradipt) [inline podvm/README.md:122]: nit: "dockerfile files" => "dockerfiles"

**Comment by wainersm** [comment]
Addressed the latest comments from @bpradipt on this update.

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @wainersm

**Pull request merged**